dd-sdk-android/src/test/kotlin/com/datadog/android/domain/FilePersistenceStrategyTest.kt

```diff
@@ -0,0 +1,334 @@
+import com.datadog.tools.unit.invokeMethod
+import com.google.gson.JsonObject
+import com.nhaarman.mockitokotlin2.any
+import com.nhaarman.mockitokotlin2.doAnswer
+import com.nhaarman.mockitokotlin2.doReturn
+import com.nhaarman.mockitokotlin2.whenever
+import fr.xgouchet.elmyr.Forge
+import fr.xgouchet.elmyr.junit5.ForgeConfiguration
+import fr.xgouchet.elmyr.junit5.ForgeExtension
+import java.io.ByteArrayOutputStream
+import java.io.File
+import kotlin.math.min
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.AfterEach
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.extension.ExtendWith
+import org.junit.jupiter.api.extension.Extensions
+import org.junit.jupiter.api.io.TempDir
+import org.mockito.Mock
+import org.mockito.junit.jupiter.MockitoExtension
+import org.mockito.junit.jupiter.MockitoSettings
+
+@Extensions(
+    ExtendWith(MockitoExtension::class),
+    ExtendWith(ForgeExtension::class),
+    ExtendWith(ApiLevelExtension::class),
+    ExtendWith(SystemOutputExtension::class)
+)
+@ForgeConfiguration(Configurator::class)
+@MockitoSettings()
+internal abstract class FilePersistenceStrategyTest<T : Any>(
+    val dataFolderName: String,
+    val maxMessagesPerPath: Int = MAX_MESSAGES_PER_BATCH,
+    val modelClass: Class<T>
+) {
+
+    lateinit var testedWriter: Writer<T>
+    lateinit var testedReader: Reader
+    @Mock(lenient = true)
+    lateinit var mockDeferredHandler: DeferredHandler
+    @TempDir
+    lateinit var tempDir: File
+
+    lateinit var mockContext: Context
+
+    // region Setup
+
+    @BeforeEach
+    fun `set up`(forge: Forge) {
+        mockContext = mockContext()
+        whenever(mockContext.filesDir) doReturn tempDir
+        whenever(mockDeferredHandler.handle(any())) doAnswer {
+            val runnable = it.arguments[0] as Runnable
+            runnable.run()
+        }
+        Datadog.initialize(mockContext, forge.anHexadecimalString())
+        val persistingStrategy = getStrategy()
+
+        testedWriter = persistingStrategy.getWriter()
+        testedReader = persistingStrategy.getReader()
+
+        setUp(testedWriter, testedReader)
+    }
+
+    @AfterEach
+    fun `tear down`() {
+        Datadog.invokeMethod("stop")
+    }
+
+    abstract fun getStrategy(): PersistenceStrategy<T>
+
+    abstract fun setUp(writer: Writer<T>, reader: Reader)
+
+    abstract fun waitForNextBatch()
+
+    // endregion
+
+    // region Writer Tests
+
+    @Test
+    fun `writes full message as json`(forge: Forge) {
+        val fakeModel = fakeModel(forge)
+        testedWriter.write(fakeModel)
+        waitForNextBatch()
+        val batch = testedReader.readNextBatch()!!
+        val model = batch.asJsonArray[0].asJsonObject
+        assertMatches(model, fakeModel)
+    }
+
+    @Test
+    fun `writes minimal model as json`(forge: Forge) {
+        val fakeModel = fakeModel(forge)
```

Comment:
Nit picking, you don't have to rewrite your PR again but keep in mind when you revisit this class : why not create a 
```kotlin
    abstract fun minimalModel(forge: Forge): T
```

You don't use the fakeModel in the rest and it's much nicer for classes that override this class ;)